src/bootstrap/native.rs

```diff
@@ -111,6 +115,201 @@ pub fn prebuilt_llvm_config(
+        if !config.dry_run {
+            t!(fs::write(llvm_stamp, key));
+        }
+    }
+
+    // if self.program_out_of_date(self.llvm_stamp(), llvm_sha + str(llvm_assertions)):
+    //     self._download_ci_llvm(llvm_sha, llvm_assertions)
+    //     for binary in ["llvm-config", "FileCheck"]:
+    //         self.fix_bin_or_dylib(os.path.join(llvm_root, "bin", binary))
+    //     for lib in os.listdir(llvm_lib):
+    //         if lib.endswith(".so"):
+    //             self.fix_bin_or_dylib(os.path.join(llvm_lib, lib))
+    //     with output(self.llvm_stamp()) as llvm_stamp:
+    //         llvm_stamp.write(llvm_sha + str(llvm_assertions))
+}
+
+fn download_ci_llvm(builder: &Builder<'_>, llvm_sha: &str) {
+    if builder.config.dry_run {
+        return;
+    }
+
+    let llvm_assertions = builder.config.llvm_assertions;
+
+    let cache_prefix = format!("llvm-{}-{}", llvm_sha, llvm_assertions);
+    let cache_dst = builder.out.join("cache");
+    let rustc_cache = cache_dst.join(cache_prefix);
+    if !rustc_cache.exists() {
+        t!(fs::create_dir_all(&rustc_cache));
+    }
+    let base = "https://ci-artifacts.rust-lang.org";
+    let url = if llvm_assertions {
+        format!("rustc-builds-alt/{}", llvm_sha)
+    } else {
+        format!("rustc-builds/{}", llvm_sha)
+    };
+    let filename = String::from("rust-dev-nightly-") + &builder.build.build.triple + ".tar.xz";
+    let tarball = rustc_cache.join(&filename);
+    if !tarball.exists() {
+        download_component(builder, base, &format!("{}/{}", url, filename), &tarball);
+        // get(
+        //     base,
+        //     "{}/{}".format(url, filename),
+        //     tarball,
+        //     self.checksums_sha256,
+        //     verbose=self.verbose,
+        //     do_verify=False,
+        //     help_on_error=help_on_error,
+        // )
+    }
+    let llvm_root = builder.config.ci_llvm_root();
+    unpack(builder, &tarball, &llvm_root);
+
+    // unpack(tarball, tarball_suffix, self.llvm_root(),
+    //  match="rust-dev",
+    //  verbose=self.verbose)
+
+    // TODO: `fix_bin_or_dylib`
+    // let llvm_lib = builder.llvm_root().join("lib");
+}
+
+fn download_component(builder: &Builder<'_>, base: &str, url: &str, dest_path: &Path) {
+    let tempfile = t!(tempfile::NamedTempFile::new());
+    let temppath = tempfile.path().to_owned();
+    drop(tempfile);
+    let tempfile_str = temppath.to_str().expect("tempdir must be valid unicode");
+    // TODO: support `do_verify`
+    download_with_retries(builder, tempfile_str, &format!("{}/{}", base, url));
+    builder.verbose(&format!("moving {} to {}", tempfile_str, dest_path.display()));
+    t!(fs::rename(temppath, dest_path));
+}
+
+fn download_with_retries(builder: &Builder<'_>, tempdir: &str, url: &str) {
+    println!("downloading {}", url);
+
+    // TODO: this check is probably not correct
+    if builder.build.build.contains("windows") {
+        for _ in 0..3 {
```

Comment:
I intentionally changed this to only be for powershell, since we already tell curl to retry. Happy to use the original logic if you want to avoid unnecessary changes.